### DISCUSSION TITLE
Components aren't rendered properly and seem to fragment into eachother

### DISCUSSION THREAD

**Discussion by RubenNijhuis** [gatsbyjs/gatsby#35593] [Help] [answered]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

When rendering the project section on my website it seems like the URLs and images are switched while the text is correct. I have been tracking down what is happening and can't explain what is going wrong. I am not "reversing" the grid or anything like that. And am just rendering the components as they come. 

My goal is to render the components correctly, with all the data that's part of one component.

If you take a look at my website you will notice that the URL that is shown isn't the same as the name of the project that you click on.

The weirdest thing about this is that after a you refresh or go back to the home page. The grid renders correctly, without fail it seems.

I tested if animations (framer-motion) were the issue but turning the `animate` prop in my project component didn't seem to fix anything.

### Reproduction Link

https://rubennijhuis.com

### Steps to Reproduce

1. Visit the my website -> https://rubennijhuis.com
2. Go the the "projects" section
3. Click on a link and compare the name of the project vs the URL

### Expected Result

I expected to go to the page I clicked on

### Actual Result

Went to an incorrect link

These images should be switched
<img width="1025" alt="image" src="https://user-images.githubusercontent.com/30152747/166902816-d406a5e5-0904-473d-adb6-32f9102a057f.png">

### Environment

```shell
System:
    OS: macOS 12.3.1
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.15.1 - /usr/local/bin/node
    Yarn: 1.22.10 - ~/.npm-global/bin/yarn
    npm: 8.4.0 - ~/.npm-global/bin/npm
  Browsers:
    Chrome: 100.0.4896.127
    Firefox: 96.0.2
    Safari: 15.4
  npmPackages:
    gatsby: ^4.2.0 => 4.11.1 
    gatsby-plugin-eslint: ^4.0.2 => 4.0.2 
    gatsby-plugin-exclude: ^1.0.2 => 1.0.2 
    gatsby-plugin-google-gtag: ^4.11.0 => 4.11.0 
    gatsby-plugin-image: ^2.4.0 => 2.11.1 
    gatsby-plugin-react-helmet: ^5.3.0 => 5.11.0 
    gatsby-plugin-robots-txt: ^1.7.0 => 1.7.0 
    gatsby-plugin-sass: ^5.4.0 => 5.11.1 
    gatsby-plugin-sharp: ^4.4.0 => 4.11.1 
    gatsby-plugin-sitemap: ^5.3.0 => 5.11.1 
    gatsby-plugin-split-css: ^2.0.2 => 2.0.2 
    gatsby-source-contentful: ^7.1.0 => 7.9.1 
    gatsby-source-filesystem: ^4.4.0 => 4.11.1 
    gatsby-transformer-sharp: ^4.4.0 => 4.11.0 
  npmGlobalPackages:
    gatsby-cli: 4.5.1
```

**Comment by LekoArts** [comment]
It all works fine for me on my end when I visit the website.

**Reply by adkanojia** [reply]
I noticed the issue on website

**Reply by RubenNijhuis** [reply]
another weird behavior is this duplicate image, it might suggest two images are layered incorrectly. https://www.dropbox.com/s/3sx6g632q4gyttn/Screen%20Recording%202022-05-05%20at%2012.41.22.mov?dl=0

**Reply by RubenNijhuis** [reply]
@LekoArts would you mind looking at the issue a bit longer than just on your own version as you have apparently closed the issue, moved it to discussion and then left as "it works fine on my end"?

**Comment by adkanojia** [comment]
I've had this issue in my project as well
I was using gatsby-source-wordpress

Are you modifying your projects data after it is fetched from contentful, and then passing into further components to render project item?
I didn't find any similar issue when I was working on that and thought the issue was in my code or something.

**Reply by RubenNijhuis** [reply]
I do sort of do this, the grid component takes the data and render a "project" component card per project. However I dont change the inherent content, only reassign the names of the keys in the object.

**Reply by RubenNijhuis** [reply]
I have tested putting everything "inline" in the projects grid component but that doesn't seem to fix things either

**Reply by RubenNijhuis** [reply]
another weird behavior is this duplicate image, it might suggest two images are layered incorrectly. https://www.dropbox.com/s/3sx6g632q4gyttn/Screen%20Recording%202022-05-05%20at%2012.41.22.mov?dl=0

**Reply by adkanojia** [reply]
I'm trying to fix this one on my end. Will share any progress

**Reply by RubenNijhuis** [reply]
You will probably need my space id and such for contentful no?

**Comment by adkanojia** [comment, marked as answer]
I've produced a minimal reproduction repo with a similar issue

https://github.com/adkanojia/gatsby-issues

To check this issue

1. Clone the repo
2. run `npm install`
3. run `npm run build`
4. run `npm serve`

Open the page `/opera-carta/#extras`
![before](https://user-images.githubusercontent.com/17718474/166955690-030c2e6f-885e-4d7c-980b-9845a95d94aa.png)

Now visit any other link from list on top
and go back to 'Extras' and the images will be different like
![after](https://user-images.githubusercontent.com/17718474/166955816-2aace236-3273-4a4e-bc02-fcf638aba934.png)

Can you check if your issue is similar @RubenNijhuis ?
I'll appreciate any help on this issue...

**Reply by RubenNijhuis** [reply]
<img width="3200" alt="Screen Shot 2022-05-05 at 6 28 28 PM-min" src="https://user-images.githubusercontent.com/30152747/166969760-899f5377-4ef1-44ec-8831-bd534729adcc.png">
Already on the `extras` page it seems be buggy

**Reply by RubenNijhuis** [reply]
I don't seem to be able to get buggy images on the other pages.

**Reply by adkanojia** [reply]
Hi @RubenNijhuis 

on Extras, the placeholder image, (screenshot where all images are same) is supposed to be correct image
when it shows all different images, that's the bug

You will have better idea checking 'Desserts' section.
Everyone know what ice cream looks like :smile:

**Reply by RubenNijhuis** [reply]
I have that too. How can we fix this issue? It seems to be either wrong in the html output or incorrect caching or sorts

**Reply by adkanojia** [reply]
seems like caching issue.
Can you share your code for this component?

**Reply by RubenNijhuis** [reply]
Putting all the code here would be a bit too much I think. I can explain the structure and send links to the files. My "project card" component is part of the grid and the card also uses a "asset handler" that handles all my videos and photos that come in.

Projects grid -> https://github.com/RubenNijhuis/Portfolio22/tree/main/src/components/ProjectsGrid
Project card -> https://github.com/RubenNijhuis/Portfolio22/tree/main/src/components/Project
Asset handler -> https://github.com/RubenNijhuis/Portfolio22/tree/main/src/components/AssetHandler

Projects grid -> project cards -> asset handler

**Reply by RubenNijhuis** [reply]
I eventually got it fixed. The cause seemed to be me swapping the elements in the projects array before render. For some reason this made everything buggy

**Reply by adkanojia** [reply]
Hi @RubenNijhuis that's great.
My code has exactly same implementation. Swapping data before rendering causes the images to be rendered from initial data.
This should be treated as a bug, right?

**Reply by RubenNijhuis** [reply]
I don't know enough about the implementation of Gatsby to know if this is related to the browser or Gatsby. Someone higher up should look at this to figure out what's happening @LekoArts

**Reply by LarsOlt** [reply]
I am having the same issue, but only when serving the gatsby build. How did you avoid changing the array? I have to sort the data returned by `useStateQuery` before rendering it.

**Reply by adkanojia** [reply]
@LarsOlt try adding a `useEffect()` hook to re-render after data is sorted.

**Reply by LarsOlt** [reply]
Thanks for the help. The issue is fixed. The images are now in correct order after rerenders. 

Here is my solution:

![image](https://user-images.githubusercontent.com/32353679/176172338-613c5f8e-c6ef-424d-a908-c6a4ef8e38d9.png)
